### PR TITLE
MM-22249 : Add automatic lazy loading of syntax highlighting libraries

### PR DESCRIPTION
**Pull request by sk409**
#### Summary

Load language libraries of highlight.js on demands to reduce bundle size.

#### Ticket Link

Fixes https://github.com/mattermost/mattermost-server/issues/14029
Jira: https://mattermost.atlassian.net/browse/MM-22249

#### Related Pull Requests

No related PRs.

#### Screenshots

No UI changes.

#### Release Note

```release-note
NONE
```

### CODE REVIEW

**Comment by neallred** [comment]
Nice! Tested this out on the spinwick and can see the languages loading only once they are needed. Can also confirm the languages are split out into separate bundles.

language-bundles

**Review by neallred — APPROVED** [review]

File: components/code_block/__snapshots__/code_block.test.tsx.snap [hunk 1]
```diff
@@ -115,9 +115,9 @@ exports[`codeBlock should render html code block with proper indentation 1`] = `
         dangerouslySetInnerHTML={
           Object {
             "__html": "\`\`\`html
-<span class=\\"hljs-tag\\">&lt;<span class=\\"hljs-name\\">div</span> <span class=\\"hljs-attr\\">className</span>=<span class=\\"hljs-string\\">&#x27;myClass&#x27;</span>&gt;</span>
-  <span class=\\"hljs-tag\\">&lt;<span class=\\"hljs-name\\">a</span> <span class=\\"hljs-attr\\">href</span>=<span class=\\"hljs-string\\">&#x27;https://randomgibberishurl.com&#x27;</span>&gt;</span>ClickMe<span class=\\"hljs-tag\\">&lt;/<span class=\\"hljs-name\\">a</span>&gt;</span>
-<span class=\\"hljs-tag\\">&lt;/<span class=\\"hljs-name\\">div</span>&gt;</span>
+&lt;div className=&apos;myClass&apos;&gt;
```
Comment (hmhealey) [inline components/code_block/__snapshots__/code_block.test.tsx.snap:118]: Would it be possible to add a test to `code_block.test.tsx` that renders these both before and after loading the library? Ideally, this snapshot would look the same once the library is loaded.

File: utils/syntax_highlighting.tsx [hunk 2]
```diff
@@ -204,3 +80,203 @@ function getLanguageFromNameOrAlias(name: string) {
         return aliases && aliases.find((a) => a === langName);
     });
 }
+
+async function registerLanguage(languageName: string) {
+    let language: LanguageFn | undefined;
+    switch (languageName) {
```
Comment (hmhealey) [inline utils/syntax_highlighting.tsx:86]: Suggestion for structuring this a bit differently, I don't think we can construct the import path dynamically (I feel like I've tried that before unsuccessfully), but could we convert this into a dictionary to reduce the number of lines here and get rid of all the break statements?

I think something like this should help keep the number of lines down while still being fairly readable as long as you make sure to not run the 

```typescript
const languageImports = {
    '1c': () => import('highlight.js/lib/languages/1c'),
    actionscript: () => import('highlight.js/lib/languages/actionscript'),
    // ...
};

if (!languageImports[languageName]) {
    return;
}

const language = (await langaugeImports[languageName]()).default;
```

**Review by hmhealey — CHANGES_REQUESTED** [review]
Thanks for the PR! I'm really excited to have this in. There's just a couple small things I'd like looked at

File: utils/syntax_highlighting.tsx (hunk 2 above)
Comment (sk409) [reply utils/syntax_highlighting.tsx]: I used map instead of switch.

File: components/code_block/__snapshots__/code_block.test.tsx.snap (hunk 1 above)
Comment (sk409) [reply components/code_block/__snapshots__/code_block.test.tsx.snap:473]: I added test to take snapshot before and after syntax highlighting.

**Comment by sk409** [comment]
@hmhealey 
Thank you for your review!
I fixed what you mentioned.

**Review by hmhealey — APPROVED** [review]
Awesome, thanks so much. I'm really excited to get this one in

**Comment by jgilliam17** [comment]
@sk409 can you please update so I can trigger e2es. Thanks 🙂  
last [cycle](https://automation-dashboard.vercel.app/cycles/a2bdf4df-93ae-4375-84aa-2aa6eedb52ca) failed to run.

**Comment by sk409** [comment]
@jgilliam17 
I commented on another PR, but just in case, What does /update-branch do?
Should I merge master into this branch?

**Comment by neallred** [comment]
@sk409 In this case, you should merge master into this branch and resolve the conflicts. `/update-branch` tries to merge in master, but it aborts if there are any conflicts.

**Comment by hmhealey** [comment]
Heh, I like how both of you commenting that still triggered the bot, even when it was in a code block. That seems like it shouldn't work that way.

But yeah, if you could please merge the latest version of master into this and resolve the conflicts, that would unblock testing

**Comment by sk409** [comment]
I'm sorry, I've been busy lately, but I'll start working on this tonight.

**Comment by sk409** [comment]
done!

**Review by jgilliam17 — APPROVED** [review]

**Pull request merged**